File: crates/flexiq-server/src/grpc/producer/enqueue.rs
//! `Enqueue` and `EnqueueBatch`.
//!
//! Uniqueness, debouncing and batching are options and shapes here, not verbs:
//! there is one `Enqueue`, and which storage entry point it reaches is decided
//! by what the request's options carry. The SDKs expose one `enqueue` with
//! options for the same reason, and the wire should not expose the dispatch
//! inside it.

use flexiq_core::job::{now_millis, Job, NewJob};
use flexiq_core::storage::Storage;
use flexiq_core::StorageBackend;
use tonic::{Response, Status};

use super::convert::{self, Blobs};
use super::Producer;
use crate::grpc::blocking::on_storage;
use crate::grpc::pb;
use crate::grpc::status::WireError;

/// Submit one job.
pub async fn one(
    producer: &Producer,
    request: pb::EnqueueRequest,
) -> Result<Response<pb::EnqueueResponse>, Status> {
    let prepared = prepare(request, producer.namespace())?;
    let (job, deduplicated) =
        on_storage(producer.storage(), move |storage| prepared.submit(storage)).await?;

    Ok(Response::new(pb::EnqueueResponse {
        // A producer that just submitted a job already has the payload it sent,
        // and asking for it back would double every enqueue's response size.
        job: Some(convert::job_to_wire(job, Blobs::NONE)),
        deduplicated,
    }))
}

/// Submit many jobs, one result per input item in input order.
///
/// The failure shape follows the backend rather than papering over it, because
/// no promise of atomicity would hold on both:
///
/// * where a batch is one transaction, one item's failure rolls back every
///   insert, so the **RPC** fails — returning the earlier items as `enqueued`
///   would report jobs that do not exist;
/// * where a batch can partially apply, the RPC succeeds and an `error` arm
///   means that one item did not land.
///
/// A client that treats an `enqueued` arm as durable is correct under both,
/// which is the property the split exists to give it.
pub async fn batch(
    producer: &Producer,
    request: pb::EnqueueBatchRequest,
) -> Result<Response<pb::EnqueueBatchResponse>, Status> {
    if request.items.is_empty() {
        return Ok(Response::new(pb::EnqueueBatchResponse {
            results: Vec::new(),
        }));
    }

    // Request-shape failures are attributable to one item, so they name it —
    // and they are refused before anything is written, whatever the backend.
    let mut prepared = Vec::with_capacity(request.items.len());
    for (index, item) in request.items.into_iter().enumerate() {
        let item = prepare(item, producer.namespace()).map_err(|error| error.at_index(index))?;
        // Storage has no batched debounce, so honouring one here would mean
        // leaving the batch to submit the rest — which silently costs the
        // transactional backends the atomicity that is the reason to send a
        // batch at all. Refusing says so instead of trading it away quietly.
        if matches!(item.dispatch, Dispatch::Debounced(_)) {
            return Err(WireError::invalid_request(
                "options.debounce is not available in a batch; enqueue a debounced job on its own",
            )
            .at_index(index)
            .into());
        }
        prepared.push(item);
    }

    let atomic = batch_is_atomic(producer.storage());
    let results = on_storage(producer.storage(), move |storage| {
        Ok(if atomic {
            submit_atomically(storage, prepared)
        } else {
            submit_one_at_a_time(storage, prepared)
        })
    })
    .await?;

    match results {
        Ok(results) => Ok(Response::new(pb::EnqueueBatchResponse { results })),
        // The whole batch rolled back. The error keeps the reason storage
        // raised; it carries no `index`, because nothing storage returns
        // attributes an all-or-nothing failure to one item, and inventing a
        // position would be worse than admitting there is none.
        Err(error) => Err(error.into()),
    }
}

/// Whether this backend's batch enqueue is one transaction.
///
/// The Diesel backends run dependency validation and every chunked insert
/// inside one `write_transaction`. Redis issues its writes as a pipeline, which
/// travels in one round trip and rolls nothing back — and its unique variant
/// does not even pipeline, it loops the single-job path.
fn batch_is_atomic(storage: &StorageBackend) -> bool {
    match storage {
        StorageBackend::Sqlite(_) => true,
        #[cfg(feature = "postgres")]
        StorageBackend::Postgres(_) => true,
        #[cfg(feature = "redis")]
        StorageBackend::Redis(_) => false,
    }
}

fn submit_atomically(
    storage: &StorageBackend,
    prepared: Vec<Prepared>,
) -> Result<Vec<pb::EnqueueBatchItemResult>, WireError> {
    // One call, so one transaction. Debounce has no batch entry point, so a
    // debounced item is refused in `prepare` rather than quietly costing the
    // rest of the batch its atomicity.
    let jobs = storage
        .enqueue_unique_batch_reporting(prepared.into_iter().map(|item| item.job).collect())
        .map_err(|error| WireError::from_queue_error(&error))?;

    Ok(jobs.into_iter().map(enqueued).collect())
}

fn submit_one_at_a_time(
    storage: &StorageBackend,
    prepared: Vec<Prepared>,
) -> Result<Vec<pb::EnqueueBatchItemResult>, WireError> {
    Ok(prepared
        .into_iter()
        .enumerate()
        .map(|(index, item)| match item.submit(storage) {
            Ok(result) => enqueued(result),
            Err(error) => pb::EnqueueBatchItemResult {
                outcome: Some(pb::enqueue_batch_item_result::Outcome::Error(
                    WireError::from_queue_error(&error).at_index(index).into(),
                )),
            },
        })
        .collect())
}

fn enqueued((job, deduplicated): (Job, bool)) -> pb::EnqueueBatchItemResult {
    pb::EnqueueBatchItemResult {
        outcome: Some(pb::enqueue_batch_item_result::Outcome::Enqueued(
            pb::EnqueueResponse {
                job: Some(convert::job_to_wire(job, Blobs::NONE)),
                deduplicated,
            },
        )),
    }
}

/// A validated request, and which storage entry point it goes to.
///
/// Validation happens before the blocking hop so a malformed request never
/// occupies a pool thread, and so a batch can attribute a bad item to its
/// index before anything is written.
struct Prepared {
    job: NewJob,
    dispatch: Dispatch,
}

enum Dispatch {
    /// No key and no window: a plain insert.
    Plain,
    /// A `unique_key`, so the enqueue may return a job that was already active.
    Unique,
    /// A debounce window to open or slide.
    Debounced(flexiq_core::storage::records::DebounceOptions),
}

impl Prepared {
    fn submit(self, storage: &StorageBackend) -> flexiq_core::Result<(Job, bool)> {
        match self.dispatch {
            // A debounced enqueue can also answer with a job that already
            // existed, but that is a slid window and not a `unique_key` match,
            // and `deduplicated` says only the latter.
            Dispatch::Debounced(options) => storage
                .enqueue_debounced(self.job, options)
                .map(|job| (job, false)),
            Dispatch::Unique => storage.enqueue_unique_reporting(self.job),
            Dispatch::Plain => storage.enqueue(self.job).map(|job| (job, false)),
        }
    }
}

/// Validate one request and decide where it goes.
fn prepare(request: pb::EnqueueRequest, namespace: &str) -> Result<Prepared, WireError> {
    // An absent body is not an empty one: `raw = ""` is a zero-length payload,
    // and no arm at all is a request that forgot to say what to run.
    let payload = match request.body {
        Some(pb::enqueue_request::Body::Raw(bytes)) => bytes,
        None => {
            return Err(WireError::invalid_request(
                "no body arm is set; send raw = \"\" for a job with no payload",
            ))
        }
    };

    let options = request.options.unwrap_or_default();
    let debounce = convert::debounce_options(&options)?;
    let has_unique_key = options.unique_key.is_some();

    let dispatch = match debounce {
        Some(options) => Dispatch::Debounced(options),
        None if has_unique_key => Dispatch::Unique,
        None => Dispatch::Plain,
    };

    let job = convert::new_job(
        request.task_name,
        payload,
        Some(options),
        namespace,
        now_millis(),
    )?;

    Ok(Prepared { job, dispatch })
}

File: sdks/go/enqueue.go
package flexiq

import (
	"context"
	"fmt"
	"time"

	pb "github.com/ByteVeda/flexiq/sdks/go/v2/internal/pb/flexiq/v1"
	"google.golang.org/grpc/status"
)

// EnqueueRequest is one job to submit.
//
// The body is Args and Kwargs, encoded into the cross-SDK envelope before the
// call goes out. Raw replaces both for a caller that already has an encoded
// envelope — a payload read off another job, or one built by hand.
type EnqueueRequest struct {
	// Task is the task to run. It is not validated here: the server holds no
	// task registry, so enqueuing a name nobody implements succeeds and the
	// job dead-letters later.
	Task string
	// Args are the positional arguments.
	Args []any
	// Kwargs are the keyword arguments, for tasks written in a language that
	// has them. The envelope carries the map either way.
	Kwargs map[string]any
	// Raw is a pre-encoded payload envelope. When set, Args and Kwargs are
	// ignored and these bytes reach storage untouched.
	Raw []byte
	// Options are the job's knobs. The zero value is a job on the default
	// queue at the server's defaults.
	Options EnqueueOptions
}

// EnqueueOptions are the producer-settable knobs on a job.
//
// Every field's zero value means "the server's default", so a caller sets only
// what it cares about. There is deliberately no namespace: a namespace is a
// property of the credential, and anything a client could name is something a
// client could forge.
type EnqueueOptions struct {
	// Queue is empty for the default queue.
	Queue string
	// Priority: higher runs first.
	Priority int32
	// MaxRetries is attempts after the first. Zero means the job is not
	// retried.
	MaxRetries int32
	// ScheduledAt is when the job becomes eligible to run. Zero means as soon
	// as possible.
	ScheduledAt time.Time
	// Timeout is how long one attempt may run. Zero takes the server's
	// default.
	Timeout time.Duration
	// UniqueKey dedupes against the *active* job carrying the same key.
	//
	// It is not an idempotency key: once the original completes or
	// dead-letters the key is released, and the same request enqueues a second
	// job. A retry loop's total deadline belongs inside the job's own life.
	UniqueKey string
	// Metadata is opaque JSON text, byte-preserved.
	Metadata string
	Notes    string
	// DependsOn are job ids this job waits on. An id in another namespace is
	// refused exactly as a missing one is.
	DependsOn []string
	// ExpiresAt: after this instant the job is cancelled instead of
	// dispatched.
	ExpiresAt time.Time
	// ResultTTL is how long to keep the result after completion.
	ResultTTL time.Duration
	// Debounce collapses a burst of enqueues carrying one key into a single
	// run. Nil for no debouncing.
	Debounce *Debounce
}

// Debounce collapses a burst of enqueues carrying one key into a single run.
//
// While a job with this key is still pending and unclaimed, its scheduled time
// slides forward instead of a second job being inserted.
type Debounce struct {
	// Key is non-empty. Jobs sharing it share a window.
	Key string
	// Window is how far ahead of now each enqueue pushes the run. Positive.
	Window time.Duration
	// MaxWait is the ceiling on the total delay, measured from the pending
	// job's creation. Mandatory, and at least Window: without it a producer
	// that never stops enqueuing starves the job forever.
	MaxWait time.Duration
	// ReplacePayload overwrites the pending job's payload with this one. False
	// keeps the payload the window opened with.
	ReplacePayload bool
	// MaxPending is the producer's own admission cap for the target queue. Nil
	// means uncapped, and then nothing is counted.
	MaxPending *int64
}

// EnqueueResult is one submitted job.
type EnqueueResult struct {
	Job Job
	// Deduplicated means the unique key matched a job that was already active,
	// and Job is that job rather than a new one.
	//
	// It is the one response field that describes what the call did rather
	// than the state it left behind, because it is the whole reason a producer
	// sets a unique key: without it the two calls are indistinguishable.
	Deduplicated bool
}

// BatchResult is one item's outcome from [Client.EnqueueBatch], in the position
// the item was sent in. Exactly one of Result and Err is set.
type BatchResult struct {
	Result *EnqueueResult
	// Err is an [*Error] describing why this item alone did not land.
	Err error
}

// Enqueue submits one job.
//
// A failure is not a signal to try again blindly. UNAVAILABLE,
// DEADLINE_EXCEEDED and CANCELLED may each mean the write landed and the
// connection dropped afterwards, and no field on the wire distinguishes them.
// A caller that retries sets [EnqueueOptions.UniqueKey] and reuses the same
// value.
func (c *Client) Enqueue(ctx context.Context, req EnqueueRequest) (EnqueueResult, error) {
	msg, err := req.toProto()
	if err != nil {
		return EnqueueResult{}, err
	}

	resp, err := c.producer.Enqueue(ctx, msg)
	if err != nil {
		return EnqueueResult{}, fromRPC(err)
	}
	return enqueueResultFromProto(resp), nil
}

// EnqueueBatch submits many jobs and answers one result per item, in input
// order.
//
// No atomicity is promised, and the call fails in two shapes a caller handles
// separately:
//
//   - Where a batch is one transaction, an item failure fails the whole RPC —
//     returning the earlier items as enqueued would report jobs that do not
//     exist. The returned error is an [*Error] whose [Error.BatchIndex] names
//     the item, and the results slice is nil.
//   - Where a batch can partially apply, the call succeeds and an item's Err
//     means that item alone did not land.
//
// Under both shapes a result with no Err is durable.
func (c *Client) EnqueueBatch(ctx context.Context, reqs []EnqueueRequest) ([]BatchResult, error) {
	items := make([]*pb.EnqueueRequest, 0, len(reqs))
	for i, req := range reqs {
		msg, err := req.toProto()
		if err != nil {
			return nil, fmt.Errorf("flexiq: batch item %d: %w", i, err)
		}
		items = append(items, msg)
	}

	resp, err := c.producer.EnqueueBatch(ctx, &pb.EnqueueBatchRequest{Items: items})
	if err != nil {
		return nil, fromRPC(err)
	}

	results := make([]BatchResult, 0, len(resp.GetResults()))
	for _, item := range resp.GetResults() {
		results = append(results, batchResultFromProto(item))
	}
	return results, nil
}

func batchResultFromProto(item *pb.EnqueueBatchItemResult) BatchResult {
	switch outcome := item.GetOutcome().(type) {
	case *pb.EnqueueBatchItemResult_Enqueued:
		result := enqueueResultFromProto(outcome.Enqueued)
		return BatchResult{Result: &result}
	case *pb.EnqueueBatchItemResult_Error:
		return BatchResult{Err: fromStatus(status.FromProto(outcome.Error))}
	default:
		// An arm this build has no name for. Reporting it as an error is the
		// honest answer: the item's outcome is genuinely unknown here, and
		// treating it as enqueued would claim a job that may not exist.
		return BatchResult{Err: fmt.Errorf("flexiq: batch item carries an outcome this client does not recognise")}
	}
}

func enqueueResultFromProto(resp *pb.EnqueueResponse) EnqueueResult {
	return EnqueueResult{
		Job:          jobFromProto(resp.GetJob()),
		Deduplicated: resp.GetDeduplicated(),
	}
}

func (r EnqueueRequest) toProto() (*pb.EnqueueRequest, error) {
	if r.Task == "" {
		return nil, fmt.Errorf("flexiq: enqueue: task name is empty")
	}

	body, err := r.body()
	if err != nil {
		return nil, err
	}
	return &pb.EnqueueRequest{
		TaskName: r.Task,
		Body:     body,
		Options:  r.Options.toProto(),
	}, nil
}

// body picks the request's payload arm. There is always one: an absent body is
// not an empty one, and the server refuses a request that sets neither.
func (r EnqueueRequest) body() (*pb.EnqueueRequest_Raw, error) {
	if r.Raw != nil {
		return &pb.EnqueueRequest_Raw{Raw: r.Raw}, nil
	}
	encoded, err := EncodeCall(r.Args, r.Kwargs)
	if err != nil {
		return nil, err
	}
	return &pb.EnqueueRequest_Raw{Raw: encoded}, nil
}

func (o EnqueueOptions) toProto() *pb.EnqueueOptions {
	opts := &pb.EnqueueOptions{
		Queue:       o.Queue,
		Priority:    o.Priority,
		MaxRetries:  o.MaxRetries,
		ScheduledAt: timestampOf(o.ScheduledAt),
		Timeout:     durationOf(o.Timeout),
		DependsOn:   o.DependsOn,
		ExpiresAt:   timestampOf(o.ExpiresAt),
		ResultTtl:   durationOf(o.ResultTTL),
		Debounce:    o.Debounce.toProto(),
	}
	if o.UniqueKey != "" {
		opts.UniqueKey = &o.UniqueKey
	}
	if o.Metadata != "" {
		opts.Metadata = &o.Metadata
	}
	if o.Notes != "" {
		opts.Notes = &o.Notes
	}
	return opts
}

func (d *Debounce) toProto() *pb.Debounce {
	if d == nil {
		return nil
	}
	return &pb.Debounce{
		Key:            d.Key,
		Window:         durationOf(d.Window),
		MaxWait:        durationOf(d.MaxWait),
		ReplacePayload: d.ReplacePayload,
		MaxPending:     d.MaxPending,
	}
}
